### ISSUE TITLE
system.activationScripts.etc assumes symlinks are files, but apparently they need not be

### ISSUE THREAD

**Issue by Smaug123** [LnL7/nix-darwin#567]
https://github.com/LnL7/nix-darwin/blob/4182ad42d5fb5001adb1f61bec3a04fae0eecb95/modules/system/etc.nix#L60 appears to assume that every symlink in `/etc/static` is a file, with the result that it blows up with exit code 1 during my `darwin-rebuild switch`: 

```
> darwin-rebuild switch --flake ~/.nixpkgs/#patrick
warning: Git tree '/Users/patrick/.nixpkgs' is dirty
building the system configuration...
warning: Git tree '/Users/patrick/.nixpkgs' is dirty
Password:
user defaults...
setting up user launchd services...
setting up /Applications/Nix Apps...
setting up pam...
applying patches...
setting up /etc...
grep: /etc/profiles/per-user/patrick: Is a directory
shasum: /etc/profiles/per-user/patrick: Is a directory  
```

The contents of my `/etc/static`:

```
> find /etc/static/
/etc/static/
/etc/static//ssl
/etc/static//ssl/certs
/etc/static//ssl/certs/ca-certificates.crt
/etc/static//nix
/etc/static//nix/registry.json
/etc/static//nix/nix.conf
/etc/static//bashrc
/etc/static//zshrc
/etc/static//ssh
/etc/static//ssh/ssh_known_hosts
/etc/static//zshenv
/etc/static//profiles
/etc/static//profiles/per-user
/etc/static//profiles/per-user/patrick
/etc/static//zprofile    
```
The directory in question:
```
> ls -la $(readlink /etc/static//profiles/per-user/patrick)
total 0
dr-xr-xr-x      6 root  wheel       192 Jan  1  1970 .
drwxrwxr-t  65535 root  nixbld  3875456 Nov 16 21:56 ..
dr-xr-xr-x    217 root  wheel      6944 Jan  1  1970 bin
dr-xr-xr-x      4 root  wheel       128 Jan  1  1970 etc
dr-xr-xr-x      2 root  wheel        64 Jan  1  1970 info
dr-xr-xr-x      8 root  wheel       256 Jan  1  1970 share   
```

For various reasons, my home-manager and/or nix-darwin setups have been broken for months, so I don't have a concrete point at which my setup broke, but for what it's worth I'm trying to build the flake in https://github.com/Smaug123/nix-dotfiles/pull/10/commits/d4f3d65405a59c1879078b85192e37f7ca641bc3, specifically the Darwin configuration (https://github.com/Smaug123/nix-dotfiles/blob/d4f3d65405a59c1879078b85192e37f7ca641bc3/flake.nix#L55). This may well turn out to be a home-manager problem, since everything was working before https://github.com/nix-community/home-manager/pull/3233, but I don't see what justifies the assertion that every symlink in `/etc/static` must be a file. 

---

The relevant snippet of the shell script being run during a `darwin-rebuild switch --flake`:

```
declare -A etcSha256Hashes
etcSha256Hashes['/etc/bashrc']='444c716ac2ccd9e1e3347858cb08a00d2ea38e8c12fdc5798380dc261e32e9ef 617b39e36fa69270ddbee19ddc072497dbe7ead840cbd442d9f7c22924f116f4 6be16cf7c24a3c6f7ae535c913347a3be39508b3426f5ecd413e636e21031e66'
etcSha256Hashes['/etc/nix/nix.conf']='7c2d80499b39256b03ee9abd3d6258343718306aca8d472c26ac32c9b0949093 19299897fa312d9d32b3c968c2872dd143085aa727140cec51f57c59083e93b9 c4ecc3d541c163c8fcc954ccae6b8cab28c973dc283fea5995c69aaabcdf785f'
etcSha256Hashes['/etc/nix/registry.json']=''
etcSha256Hashes['/etc/profiles/per-user/patrick']=''
etcSha256Hashes['/etc/ssh/ssh_known_hosts']=''
etcSha256Hashes['/etc/ssl/certs/ca-certificates.crt']=''
etcSha256Hashes['/etc/zprofile']='db8422f92d8cff684e418f2dcffbb98c10fe544b5e8cd588b2009c7fa89559c5 0235d3c1b6cf21e7043fbc98e239ee4bc648048aafaf6be1a94a576300584ef2'
etcSha256Hashes['/etc/zshenv']=''
etcSha256Hashes['/etc/zshrc']='19a2d673ffd47b8bed71c5218ff6617dfc5e8533b240b9ba79142a45f8823c23 fb5827cb4712b7e7932d438067ec4852c8955a9ff0f55e282473684623ebdfa1 c5a00c072c920f46216454978c44df044b2ec6d03409dc492c7bdcd92c94a110 40b0d8751adae5b0100a4f863be5b75613a49f62706427e92604f7e04d2e2261'

for f in $(find /etc/static/* -type l); do
  l=/etc/${f#/etc/static/}
  d=${l%/*}
  if [ ! -e "$d" ]; then
    echo mkdir -p "$d"
  fi
  if [ -e "$l" ]; then
    if [ "$(readlink "$l")" != "$f" ]; then
      echo "grepping: $l"
      if ! grep -q /etc/static "$l"; then
        o=$(shasum -a256 "$l")
        o=${o%% *}
        for h in ${etcSha256Hashes["$l"]}; do
          if [ "$o" = "$h" ]; then
            echo mv "$l" "$l.orig"
            echo ln -s "$f" "$l"
            break
          else
            h=
          fi
        done

        if [ -z "$h" ]; then
          echo "[1;31merror: not linking environment.etc.\"${l#/etc/}\" because $l already exists, skipping...[0m" >&2
          echo "[1;31mexisting file has unknown content $o, move and activate again to apply[0m" >&2
        fi
      fi
    fi
  else
    echo ln -s "$f" "$l"
  fi
done
```

**Comment by Enzime** [comment]
I believe this should be fixed by #731

**Comment by emilazy** [comment]
I think so. Please let us know if you still run into this.
